File: exercise/17/solution.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
		vector<string> keyboard = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
		vector<string> result[2];
        if (digits.length() == 0) {
			return result[0];
		}
		int now = 0;
		result[1].push_back("");
		for (int i = 0; i < digits.length(); i++) {
			int digit = digits[i] - '0';
			for (int j = 0; j < result[1-now].size(); j++) {
				for (int k = 0; k < keyboard[digit].length(); k++) {
					result[now].push_back(result[1-now][j] + keyboard[digit][k]);
				}
			}
			now = 1 - now;
			result[now].clear();
		}
		return result[1 - now];
    }
};

File: exercise/17/solution.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        phone = {'2': "abc", '3': "def", '4': "ghi", '5': "jkl", '6': "mno", '7': "pqrs", '8': "tuv", '9': "wxyz"}
        result = [""]
        for c in digits:
            new_result = []
            for prefix in result:
                for next_c in phone[c]:
                    new_result.append(prefix + next_c)
            result = new_result
        return result

